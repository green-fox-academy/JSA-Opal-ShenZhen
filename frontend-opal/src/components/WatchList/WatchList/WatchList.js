import React, { useEffect } from 'react';
import { View, SafeAreaView, ScrollView } from 'react-native';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { Accordion } from 'native-base';

import WatchListItem from 'components/WatchList/WatchListItem/WatchListItem';
import thunks from 'thunks/watchlists';
import actions from 'actions/watchlists';

const WatchList = ({ watchlists, getLists, closeInstrumentsInList }) => {
  useEffect(() => {
    getLists();
  }, [getLists]);

  const dataArray = watchlists.map(list => ({
    ...list,
    title: `Watchlist ${list.index + 1}`
  }));

  return (
    <SafeAreaView>
      <ScrollView>
        <View>
          <Accordion
            dataArray={dataArray}
            renderContent={WatchListItem}
            headerStyle={{ backgroundColor: 'white', fontWeight: 'bold' }}
            onAccordionClose={() => {
              watchlists.forEach(list => {
                closeInstrumentsInList(list.index);
              });
            }}
          />
        </View>
      </ScrollView>
    </SafeAreaView>
  );
};

WatchList.propTypes = {
  watchlists: PropTypes.arrayOf(PropTypes.object).isRequired,
  getLists: PropTypes.func.isRequired,
  closeInstrumentsInList: PropTypes.func.isRequired
};

const mapStateToProps = ({ watchlists = [] }) => ({ watchlists });

const mapDispatchToProps = dispatch => ({
  getLists: () => dispatch(thunks.getWatchlistData()),
  closeInstrumentsInList: index => dispatch(actions.closeInstrumentsInList(index))
});

WatchList.title = 'WATCHLIST';

export default connect(mapStateToProps, mapDispatchToProps)(WatchList);
