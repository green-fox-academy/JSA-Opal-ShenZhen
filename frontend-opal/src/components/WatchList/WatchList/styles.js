import { StyleSheet } from 'react-native';

export default StyleSheet.create({
  border: {
    borderWidth: 2,
    borderStyle: 'solid',
    borderColor: 'black',
    display: 'flex',
    flexDirection: 'row',
    justifyContent: 'space-between'
  },
  addNew: {
    position: 'absolute',
    width: 50,
    height: 50,
    right: 15,
    bottom: 15,
    backgroundColor: '#e17055',
    borderRadius: 50 / 2,
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'center'
  },
  buttonContainer: {
    display: 'flex',
    justifyContent: 'flex-end'
  },
  buttonOK: {
    backgroundColor: 'rgb(77,119,245)'
  },
  buttonText: {
    color: 'white'
  }
});
