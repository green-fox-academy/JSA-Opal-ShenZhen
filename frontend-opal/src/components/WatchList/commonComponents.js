import React from 'react';
import { View, Text, TouchableOpacity, Image } from 'react-native';
import PropTypes from 'prop-types';

export function FormatText({ content, blockStyle, textStyle = {} }) {
  return (
    <View style={blockStyle}>
      <Text style={textStyle}>{content}</Text>
    </View>
  );
}

FormatText.defaultProps = {
  content: '',
  blockStyle: {},
  textStyle: {}
};

FormatText.propTypes = {
  content: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
  /* eslint-disable react/forbid-prop-types */
  blockStyle: PropTypes.object,
  textStyle: PropTypes.object
  /* eslint-disable react/forbid-prop-types */
};

export function FormatTouchIcon({ iconStyle, iconSource, blockStyle, pressFunc = null }) {
  return (
    <TouchableOpacity style={blockStyle} onPress={pressFunc}>
      <Image style={iconStyle} source={iconSource} />
    </TouchableOpacity>
  );
}

FormatTouchIcon.defaultProps = {
  iconStyle: {},
  iconSource: '',
  blockStyle: {},
  pressFunc: PropTypes.func
};

FormatTouchIcon.propTypes = {
  iconStyle: PropTypes.object,
  iconSource: PropTypes.number,
  blockStyle: PropTypes.object,
  pressFunc: PropTypes.func
};
