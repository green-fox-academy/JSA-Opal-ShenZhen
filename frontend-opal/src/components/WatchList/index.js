import WatchListContainer from './WatchListContainer/WatchListContainer';

export default WatchListContainer;