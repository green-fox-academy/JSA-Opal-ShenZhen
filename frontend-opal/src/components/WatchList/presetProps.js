import PropTypes from 'prop-types';

const defaultQuote = {
  symbol: '',
  latestPrice: 0,
  changePercent: '',
  marketCap: '',
  week52High: 0,
  week52Low: 0,
  peRatio: 0,
  eps: '',
  divYield: '',
  chartData: []
};

const quoteTypes = PropTypes.shape({
  symbol: PropTypes.string,
  latestPrice: PropTypes.number,
  changePercent: PropTypes.string,
  marketCap: PropTypes.string,
  week52High: PropTypes.number,
  week52Low: PropTypes.number,
  peRatio: PropTypes.number,
  eps: PropTypes.string,
  divYield: PropTypes.string,
  chartData: PropTypes.arrayOf(
    PropTypes.shape({
      x: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
      y: PropTypes.number
    })
  )
});

export default {
  defaultQuote,
  quoteTypes
};
