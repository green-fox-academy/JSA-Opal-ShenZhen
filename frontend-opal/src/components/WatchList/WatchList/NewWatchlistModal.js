import React, { useState } from 'react';
import { View, SafeAreaView, ScrollView } from 'react-native';
import { Fab, Icon, Text, Card, CardItem, Item, Input, Label, Button, Form } from 'native-base';
import Modal from 'react-native-modal';

import styles from './styles';

const NewWatchlistModal = () => {
  const [isModalVisible, toggleModal] = useState(false);

  return (
    <SafeAreaView style={{ width: '100%', height: '100%' }}>
      <View>
        <Fab onPress={() => toggleModal(!isModalVisible)} style={{ backgroundColor: '#e17055' }}>
          <Icon type="FontAwesome" name="plus" />
        </Fab>
      </View>
      <Modal
        isVisible={isModalVisible}
        onBackdropPress={() => toggleModal(false)}
        backdropTransitionOutTiming={0}
      >
        <View>
          <Card>
            <CardItem>
              <Text>Add a new watchlist</Text>
            </CardItem>
            <CardItem>
              <Form style={{ width: '100%' }}>
                <Item underline stackedLabel>
                  <Label>Name of the watchlist</Label>
                  <Input />
                </Item>
              </Form>
            </CardItem>
            <CardItem style={styles.buttonContainer}>
              <Button style={styles.buttonOK} onPress={() => toggleModal(!isModalVisible)}>
                <Text style={styles.buttonText}>OK</Text>
              </Button>
            </CardItem>
          </Card>
        </View>
      </Modal>
    </SafeAreaView>
  );
};

export default NewWatchlistModal;
