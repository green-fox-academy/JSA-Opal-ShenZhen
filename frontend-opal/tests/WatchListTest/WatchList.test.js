import React from 'react';
import renderer from 'react-test-renderer';

import WatchList from '../../components/WatchList/WatchList';

describe('<WatchList />', () => {
  it('renders correctly', () => {
    const info = [{
      name: 'MSFT', num: '141.21', increase: '+0.23', whole: '1.2M',
    },
    {
      name: 'AMZN', num: '2108.2', increase: '-0.06', whole: '609K',
    },
    {
      name: 'AAPL', num: '250.7', increase: '+1.63', whole: '810K',
    }];

    const chartData = [
      { x: 1, y: 8 },
      { x: 2, y: 7 },
      { x: 3, y: 5 },
    ];

    const stockData = {
      '52 wk low': '231.6',
      '52 wk high': '250.7',
      'P/E': '20.3',
      EPS: '1.2',
      DivYield: '2.34%',
    };

    const watchlistArr = ['my watchlist1', 'my watchlist2'];
    const tree = renderer.create(<WatchList
      watchlistArr={watchlistArr}
      info={info}
      stockData={stockData}
      chartData={chartData}
    />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
