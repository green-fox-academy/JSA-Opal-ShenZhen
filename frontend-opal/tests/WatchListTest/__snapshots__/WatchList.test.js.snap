// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<WatchList /> renders correctly 1`] = `
<View>
  <View>
    <View
      style={
        Object {
          "borderColor": "black",
          "borderStyle": "solid",
          "borderWidth": 2,
          "display": "flex",
          "flexDirection": "row",
          "justifyContent": "space-between",
        }
      }
    >
      <Text>
        my watchlist1
      </Text>
      <View
        accessible={true}
        isTVSelectable={true}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#007aff",
            "borderBottomWidth": null,
            "borderColor": "#007aff",
            "borderLeftWidth": null,
            "borderRadius": 5,
            "borderRightWidth": null,
            "borderTopWidth": null,
            "borderWidth": undefined,
            "elevation": 2,
            "flexDirection": "row",
            "height": 45,
            "justifyContent": "space-between",
            "opacity": 1,
            "paddingBottom": 6,
            "paddingTop": 6,
            "shadowColor": undefined,
            "shadowOffset": undefined,
            "shadowOpacity": undefined,
            "shadowRadius": undefined,
          }
        }
      >
        <Text />
      </View>
    </View>
  </View>
  <View>
    <View
      style={
        Object {
          "borderColor": "black",
          "borderStyle": "solid",
          "borderWidth": 2,
          "display": "flex",
          "flexDirection": "row",
          "justifyContent": "space-between",
        }
      }
    >
      <Text>
        my watchlist2
      </Text>
      <View
        accessible={true}
        isTVSelectable={true}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#007aff",
            "borderBottomWidth": null,
            "borderColor": "#007aff",
            "borderLeftWidth": null,
            "borderRadius": 5,
            "borderRightWidth": null,
            "borderTopWidth": null,
            "borderWidth": undefined,
            "elevation": 2,
            "flexDirection": "row",
            "height": 45,
            "justifyContent": "space-between",
            "opacity": 1,
            "paddingBottom": 6,
            "paddingTop": 6,
            "shadowColor": undefined,
            "shadowOffset": undefined,
            "shadowOpacity": undefined,
            "shadowRadius": undefined,
          }
        }
      >
        <Text />
      </View>
    </View>
  </View>
</View>
`;
